### DISCUSSION TITLE
What's the most appropriate way to detect when a user is idle or not using a textual app?

### DISCUSSION THREAD

**Discussion by dontascii** [Textualize/textual#4700] [Q&A] [answered]
Hi, 
I have an app which has a worker.  I would like that worker to stop when a user is idle for a period of time or if they minimize the terminal window/switch to a different application. (and maybe start the worker back up when they return to the app.  But lets focus on stopping the worker for now) 

I can handle the minimize window/switch application case with 'Blur' (and/or `AppBlur`) but I'm not sure about detecting idle state. (no mouse or keyboard events) for a period of time. I know the message pump has a `check_idle` but the docs indicate that will generate an _idle_ event. The docs also say that the _idle_ event is a pseudo event which doesn't go through the message queue. 

So, I'm thinking I would need to implement a global `on_event` handler and record the timestamp when each event arrives.  Then I could compare that with the current timestamp and if the period of time I want has passed, then stop the worker.  If I'm not mistaken, a timer could be used for the date comparison as it does not generate any events. (Like a worker would do) 

Before I continue down this path however, I wanted to see if there is already a solution that works well for my needs and get any insight from all of you smart folk on how to implement this. 

Thanks in advance!

**Comment by willmcgugan** [comment, marked as answer]
I think you are on the right track, but I think it will be simpler that what you have in mind. Implement `on_mount`, and start or reset a timer on an inout event, for whatever period you would consider _idle_.

**Reply by dontascii** [reply]
Thanks Will. You were right.  I was missing the fact that a timer can be reset with `reset()`, which made this a lot easier than what I was originally thinking.  The timer just stops the worker after 5 min and the mouse / input events reset the timer.  Elegant and simple. 
Thanks for your help.
